File: python_sdk/src/eva_sdk/eva_client.py
import uuid
from dataclasses import dataclass
from typing import Dict, Any, TypeVar, Type
from urllib.parse import urljoin, urlparse

from postchain_client_py import BlockchainClient, SignatureProvider, NetworkSettings

@dataclass
class EngineConfig:
    url: str
    prefix: str
    pub: str

T = TypeVar('T', bound='EvaClient')

class EvaClient:
    def __init__(
        self,
        private_key: str,
        client: BlockchainClient,
        config: NetworkSettings,
        engine: EngineConfig
    ):
        self.private_key = private_key
        self.config = config
        self.client = client
        self.signature_provider = SignatureProvider(priv_key=private_key)
        
        # Ensure URL doesn't end with '/'
        url = engine.url[:-1] if engine.url.endswith('/') else engine.url
        self.engine = EngineConfig(
            url=url,
            prefix=engine.prefix,
            pub=engine.pub
        )

    @classmethod
    async def init(
        cls: Type[T],
        private_key: str,
        chain_config: Dict[str, Any],
        engine: EngineConfig
    ) -> T:
        client = await BlockchainClient.create(chain_config)
        return cls(private_key, client, chain_config, engine)

    async def get_engine(self) -> Dict[str, Any]:
        engine = await self.client.query("get_engine_by_address", {"address": self.engine.pub})
        return engine

    def _generate_request_uid(self) -> str:
        return f"{self.engine.prefix}-{str(uuid.uuid4())}"

    async def sign_evaluate_tweet_request(self, input_tweet: str, output_tweet: str) -> str:
        request_uid = self._generate_request_uid()
        operation = {
            "name": "evaluate_tweet_request",
            "args": [request_uid, input_tweet, output_tweet]
        }
        
        signed_tx = await self.client.sign_transaction(
            operations=[operation],
            signers=[
                self.signature_provider.pub_key,
                bytes.fromhex(self.engine.pub)
            ],
            signature_provider=self.signature_provider
        )
        return signed_tx.hex()

    async def submit_evaluate_tweet_request(self, tx_hash: str) -> str:
        engine_path = urlparse(self.engine.url).path
        endpoint = 'eval/evaluate-tweet-request'
        full_url = urljoin(self.engine.url, f"{engine_path}/{endpoint}")
        
        print(full_url)
        async with self.client.rest_client.session.post(
            full_url,
            json={"hash": tx_hash},
            headers={"Content-Type": "application/json"}
        ) as response:
            data = await response.text()
            print(data)
            return data

File: js_sdk/src/EvaClient.ts
import { createClient, IClient, NetworkSettings, newSignatureProvider, SignatureProvider } from "postchain-client";
import { chainConfig, CHROMIA_CHAIN } from "./config";
import { v4 as uuidv4 } from 'uuid';

interface EngineConfig {
  url: string;
  prefix: string;
  pub: string;
}

export class EvaClient {
  protected static instance: EvaClient;
  public engine: EngineConfig;
  public client: IClient;
  public signatureProvider: SignatureProvider;
  public config: NetworkSettings;

  constructor(
    protected readonly privateKey: string,
    client: IClient, 
    config: NetworkSettings,
    engine: EngineConfig
  ) {
    this.config = config;
    this.client = client;
    this.signatureProvider = newSignatureProvider({
      privKey: this.privateKey,
    });
    this.engine = {    
      url: engine.url.endsWith('/') ? engine.url.slice(0, -1) : engine.url,
      prefix: engine.prefix,
      pub: engine.pub
    }
  }

  static async init<T extends EvaClient>(
    this: new (privateKey: string, client: IClient, config: NetworkSettings, engine: EngineConfig) => T,
    privateKey: string,
    chain: CHROMIA_CHAIN,
    engine: EngineConfig
  ): Promise<T> {
    const client = await createClient({
      ...chainConfig[chain],
    });
    const instance = new this(privateKey, client, chainConfig[chain], engine);
    return instance as T;
  }

  async getEngine() {
    const engine = await this.client.query("get_engine_by_address", { address: this.engine.pub });
    return engine;
  }

  private generateRequestUid(): string {
    return this.engine.prefix + "-" + uuidv4();
  }

  async signEvaluateTweetRequest(inputTweet: string, outputTweet: string): Promise<string> {
    const requestUid = this.generateRequestUid();
    const operation = {
        name: "evaluate_tweet_request",
        args: [requestUid, inputTweet, outputTweet]
    };
    const signedTx = await this.client.signTransaction({
        operations: [
            operation
        ],
        signers: [
            this.signatureProvider.pubKey,
            Buffer.from(this.engine.pub, "hex")
        ]
    }, this.signatureProvider);
    return signedTx.toString("hex");
  }

  async submitEvaluateTweetRequest(txHash: string): Promise<{
    result: any,
    tx: string
  }> {
    const engineUrl = new URL(this.engine.url).pathname + '/eval/evaluate-tweet-request';
    const fullUrl = new URL(engineUrl, this.engine.url).toString();
    const response = await fetch(fullUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ hash: txHash })
    });
    const data = await response.json();
    return data;

  }
}
